docs/proposals/fast-sync/trust.md

```diff
@@ -0,0 +1,499 @@
+   big. It's not yet clear how much time it will take in practice, but
+   probably not just few seconds. If it takes a minute and `l = k`, first three
+   slots among last `2 · k` will be missed, because their leaders will
+   be computing slot leaders (because it's mandatory to put a
+   signature of this list into new block).
+2. If leaders must be computed before creating a block for `8 · k`-th
+   slot, it's enough to force rollback of only one block to require
+   nodes to recalculate slot leaders. It's not very expensive to make
+   a fork of depth 1 (or 2, 3). However, computing slot leaders is an
+   expensive operation, so we'd like to avoid doing it more than once
+   for epoch.
+
+If we pick `l` slightly smaller than `k`, both problems will be
+resolved with high probability. In Cardano SL `k = 2160`. Suppose we
+say `l = 2100`. Then:
+1. Each node will have 40 minutes to compute list of slot
+   leaders. 20 minutes should be enough for everybody.
+2. Rollback of 60-120 (depending on chain quality) blocks is very
+   unlikely to happen, so almost always it will be enough to compute
+   slot leaders only once.
+3. However, we will assume that there is at least one honest block in
+   2100 blocks, not in 2160. Probably it's not a big problem, because
+   probability of all blocks being invalid decreases exponentially, so
+   there should be significant difference between 2100 and 2160 (in
+   both cases probability is negligible).
+
+### Properties
+
+Let's assume that `l ≈ k = 2160`.
+
+#### Correctness of the snapshot
+
+The probability of the snapshot being incorrect is the same as the
+probability of a fork in Ouroboros protocol with depth more than
+`l`. If `l` is almost the same as `k`, this probability is negligible.
+
+Recall the steps from
+[fast synchronization procedure](#fast-synchronization-procedure). The
+node can easily verify tiny block for `e`, because it knows leaders
+for epoch `e` (it has all blocks before `e`). For tiny block for
+epoch `e + 1` it's only possible to verify that signatures are valid
+and correspond to actual slot leaders, but it's impossible to verify
+list of slot leaders put into that tiny block. However, it's easy to
+see that if signatures are valid, list of leaders should also be valid
+if Common Prefix property with parameter `l` is not violated. In this
+case among last `2 · l` slots of epoch `e` there are at least `l`
+blocks and at least one of them must be produced by an honest
+stakeholder according to Common Prefix property. It implies that at
+least one signature in the tiny block for epoch `e + 1` is produced by
+an honest stakeholder, so the whole list must be correct. Verification
+of subsequent tiny blocks is the same.
+
+If all tiny blocks are correct, headers received at step 8 are
+guaranteed to be created by real slot leaders. According to Common
+Prefix property, at least the oldest block must be created by an
+honest stakeholder, hence it must contain the hash of the correct
```

Comment:
Problem is it's not guaranteed by described algorithm that snapshot hash will appear in oldest block. By described algorithm we enforce only inclusion of leader list signatures